### ISSUE TITLE
An issue about `_gather`

### ISSUE THREAD

**Issue by JDYG** [hamarh/HMNet_pth#2]
Hi, 
The `_gather()` function in `HMNet.py` seems to have two `return` ([return list_dst](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L177) and [return outputs, states](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L185)). Is it correct for this function?

Besides, in the [forward function](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L134C52-L134C52), the `_gather` should have 3 return values (`outputs1, outputs2, outputs3`). However, no matter whether it's `return list_dst` or `return outputs, states`, neither matches the expected number of return values. I am confused about this function, could you please check it or give some explanation about the `_gather` function？

**Comment by hamarh** [comment]
Hi,
Thanks for pointing out. The second return ([return outputs, states](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L185)) is the old code which is not used now and the first return ([return list_dst](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L177)) is correct. Now, I have removed the unused return.

> Besides, in the [forward function](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L134C52-L134C52), the _gather should have 3 return values (outputs1, outputs2, outputs3). However, no matter whether it's return list_dst or return outputs, states, neither matches the expected number of return values. I am confused about this function, could you please check it or give some explanation about the _gather function？

In the [forward function](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L134C52-L134C52), the `_gather` function has 3 return values; The `_gather` function takes `(output1, output2, output3)` as the first argument, `list_dst`, updates it with `(out1, out2, out3)` based on `gather_indices` and `time_idx`, and returns updated `list_dst`, i.e. `(output1, output2, output3)`.

**Comment by JDYG** [comment]
Hi, thanks for your response.
So if I delete the following code in the `_gather()` function in `hmnet.py`
```python
        # idx_offset += len(list_keys)
        #
        # states = (idx_offset, z1, z2, z3)
        # outputs = (out_z1, out_z2, out_z3)
        #
        # return outputs, states
```

The `idx_offset` will not update during the whole training process. It means that the [logical and ](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L260C16-L260C16) to select the mask index will always be False (in the `_identify_required_outputs` in hmdet.py).  This will cause the output of `_identify_required_outputs_batch` be empty.

I use your default parameters (`TRAIN_DURATION = 200e3
DELTA_T = 5e3
INPUT_SIZE = (240, 304)`) to run the `train.py` for Gen1 detection. I found that the [gather_indices](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L100) is always empty, which makes the [features ](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L107) generated by `self.backbone` is always empty.

**Comment by JDYG** [comment]
I guess the `idx_offset += len(list_keys)` should be put before the [`return list_dst`](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/base/backbone/hmnet.py#L177C9-L177C24). Like the following code. But I am not sure.

```python
    def _gather(self, list_dst, list_src, gather_indices, time_idx):
        time_indices = gather_indices['time']
        batch_indices = gather_indices['batch']
        assert len(time_indices) == len(batch_indices)
        destination = torch.arange(len(time_indices))

        mask = time_indices == time_idx
        src_indices = batch_indices[mask]
        dst_indices = destination[mask]

        for dst, src in zip(list_dst, list_src):
            dst[dst_indices] = src[src_indices]

        idx_offset += len(list_keys)
        return list_dst

 #       states = (idx_offset, z1, z2, z3)
 #       outputs = (out_z1, out_z2, out_z3)
 #
 #       return outputs, states
```

**Comment by JDYG** [comment]
After comparing the [`hmdepth.py` for depth](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/depth/hmdepth.py#L75C1-L99C23) and `hmdet.py` for detection. I guess the `self.idx_offset += len(list_events)` should be added before the `return outputs`. But I am not sure. 


```python
# the forward function in hmdet.py
    def forward(self, list_events, list_image_metas, list_gt_bboxes, list_gt_labels, list_ignore_masks, init_states=True) -> Tensor:
        if init_states:
            self.idx_offset = 0

        # Gather gts for loss calculation
        gather_indices = self._identify_required_outputs_batch(list_gt_labels, self.idx_offset)
        print(gather_indices)
        out_image_metas  = self._gather(list_image_metas , gather_indices)
        out_gt_bboxes    = self._gather(list_gt_bboxes   , gather_indices)
        out_gt_labels    = self._gather(list_gt_labels   , gather_indices)
        out_ignore_masks = self._gather(list_ignore_masks, gather_indices)

        # Backbone
        features = self.backbone(list_events, list_image_metas, gather_indices, init_states=init_states, detach=True, fast_training=True)

        # Head
        loss, log_vars = self._forward_head(features, out_gt_bboxes, out_gt_labels, out_ignore_masks)

        # Make sure all parameters are involved in loss calculation
        loss = loss + sum([ 0. * params.sum() for params in self.parameters() ])

        outputs = dict(loss=loss, log_vars=log_vars, num_samples=len(out_image_metas))
        self.idx_offset += len(list_events)  # I guess the update for idx_offset should be added here
        return outputs
```

**Comment by hamarh** [comment]
Yes, you are right. There should be `self.idx_offset += len(list_events)` added at the end of forward function.

But this may not solve the problem you are facing because `idx_offset` is always zero when trained without TBPTT.
`idx_offset` is only used to store the timestep index for second forward call in TBPTT training.

>The idx_offset will not update during the whole training process. It means that the [logical and ](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L260C16-L260C16)to select the mask index will always be False (in the _identify_required_outputs in hmdet.py). This will cause the output of _identify_required_outputs_batch be empty.

The mask index should become True even when `idx_offset` is zero. The mask is computed as 
`mask = torch.logical_and(time_indices < len(list_gt), (time_indices + idx_offset) > self.backbone.warmup)`.
Then, `time_indices < len(list_gt)` is always True (we may not need this condition) and `(time_indices + idx_offset) > self.backbone.warmup)` become True when the time_indices exceeds the warmup steps.

So, I don't know why the [gather indices](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L100) become empty in your case.
Could you check if `time_indices` in [logical and](https://github.com/hamarh/HMNet_pth/blob/a9740820cfcf11867970c556efeeced92c0926e6/hmnet/models/detection/hmdet.py#L260C16-L260C16) is empty or not?

**Comment by JDYG** [comment]
Hi,
I found that the [list_gt](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/hmnet/models/detection/hmdet.py#L265) input to `_identify_required_outputs` is a list of length 40 (because `TRAIN_DURATION = 200e3` and `DELTA_T = 5e3`).

However, the `list_gt` only has the first element that is not [None], as shown below.
```
# the value for list_gt
# only the first one is tensor[0], others are [None]
[[tensor([0], device='cuda:2')], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None]]
```
Therefore, the `time_indices` extracted in [`for tidx, gt_labels in enumerate(list_gt)`](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/hmnet/models/detection/hmdet.py#L268C13-L268C13) is always be 1.
Since we didn't update the `idx_offset` (always 0), the `(time_indices + idx_offset)` will always be `(1+0)` and `(time_indices + idx_offset) > self.backbone.warmup` always be False.
Thereby, the `mask` is always `tensor([False])`, and time_indices is empty.

P.S. I use your providing code in `experiments/detection/data/gen1`to segment Gen1 dataset. I guess the [segment duration is 1000 ](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/experiments/detection/data/gen1/scripts/make_event_meta.py#L51).

Because the `_identify_required_outputs` is called in `_identify_required_outputs_batch`, I am not sure whether the size of `list_gt` is related to the batch size. I set the [batch size](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/experiments/detection/config/hmnet_B3_yolox.py#L212) is 16 in  `hmnet_B3_yolox.py`. Because I didn't use single GPU, so the real batch size is 1 according to [the code](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/experiments/detection/scripts/train.py#L556) in `train.py`

**Comment by hamarh** [comment]
Hmmm, there may be problem with data loading. The [gen1.py](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/hmnet/dataset/gen1.py#L243C1-L243C1) crops training sequence so that the labels are placed at the last 25% of the cropped sequence. But in your case, the labels are placed at the beginning.

I also run the training of HMNet-B1 with `--distributed` and real batch size = 1, but I coluldn't reproduce the error.
In my case, the [lsit_gt](https://github.com/hamarh/HMNet_pth/blob/5973a49d380f5255aaa8ac3cb797b725d140817e/hmnet/models/detection/hmdet.py#L265) have labels around the last timesteps, which is expected behavior.
```
# @GPU0
[[None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [tensor([0], device='cuda:0')], [None], [None], [None], [None], [None], [None]]
# @GPU1
[[None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [tensor([0], device='cuda:1')], [None], [None], [None], [None]]
# @GPU2
[[None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [tensor([0, 0], device='cuda:2')], [None], [None], [None], [None], [None], [None]]
# @GPU3
[[None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [None], [tensor([0], device='cuda:3')], [None], [None], [None], [None], [None], [None]]
```
The command used:
`python ./scripts/train.py ./config/hmnet_B1_yolox_debug.py --amp --distributed --overwrite`

**Comment by JDYG** [comment]
Thank you for your detailed information. I cloned your latest code, and it now works correctly. I tried to figure out the reason this afternoon but failed to find which code caused this bug. In the end, I gave up and decided to use the latest code.
